File: backend/app/schemas/trading.py
"""Pydantic schemas for Phase 8 paper trading endpoints."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, field_validator

# ── Order schemas ─────────────────────────────────────────────────────────────

class PlaceOrderRequest(BaseModel):
    signal_id: str
    side: str = "BUY"      # 'BUY' | 'SELL'
    quantity: int | None = None  # override signal's suggested_qty if provided

    @field_validator("side")
    @classmethod
    def _side_upper(cls, v: str) -> str:
        v = v.upper()
        if v not in ("BUY", "SELL"):
            raise ValueError("side must be BUY or SELL")
        return v

    @field_validator("quantity")
    @classmethod
    def _qty_positive(cls, v: int | None) -> int | None:
        if v is not None and v <= 0:
            raise ValueError("quantity must be > 0")
        return v


class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: int
    signal_id: str | None
    stock_id: int
    symbol: str = ""
    mode: str
    side: str
    order_type: str
    quantity: int
    price: Decimal | None
    status: str
    placed_at: datetime
    filled_at: datetime | None
    filled_price: Decimal | None
    filled_qty: int | None
    error_message: str | None


# ── Position schemas ───────────────────────────────────────────────────────────

class PositionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: int
    stock_id: int
    symbol: str = ""
    mode: str
    side: str
    quantity: int
    avg_entry_price: Decimal
    current_sl: Decimal | None
    current_tp: Decimal | None
    trail_state: str
    unrealized_pnl: Decimal | None   # NET of estimated round-trip costs (open)
    realized_pnl: Decimal            # NET of charges once closed
    charges: Decimal | None = None   # round-trip trading costs (None = pre-cost history)
    exit_price: Decimal | None = None    # closing fill price (None while open)
    exit_reason: str | None = None       # sl_hit | tp_hit | manual (None while open)
    current_price: Decimal | None = None  # transient live/last price (open positions)
    peak_price: Decimal | None = None    # best price seen while open (MFE)
    peak_pnl: Decimal | None = None      # GROSS peak profit (max favourable excursion)
    opened_at: datetime
    closed_at: datetime | None
    signal_id: str | None


class PositionListResponse(BaseModel):
    total: int
    positions: list[PositionOut]


# ── Profit-lock shadow comparator (read-only evidence) ─────────────────────────

class ShadowPolicyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    policy: str
    exit_price: Decimal | None
    exit_time: datetime | None
    exit_net: Decimal | None
    still_open: bool
    capture_pct: float | None


class ShadowComparisonOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    position_id: str
    symbol: str
    side: str
    quantity: int
    entry: Decimal
    original_sl: Decimal
    classification: str
    bars: int
    peak_price: Decimal | None
    peak_gross: Decimal | None
    actual_exit_price: Decimal | None
    actual_net: Decimal | None
    actual_capture_pct: float | None
    actual_exit_off_tape: bool = False
    policies: list[ShadowPolicyOut]
    note: str | None = None


class ShadowCompareResponse(BaseModel):
    total: int
    comparisons: list[ShadowComparisonOut]


class ClosePositionRequest(BaseModel):
    exit_price: Decimal | None = None  # if None, use current market price


class UpdateSlRequest(BaseModel):
    new_sl: Decimal

    @field_validator("new_sl")
    @classmethod
    def _positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("new_sl must be positive")
        return v


# ── Daily P&L / circuit breaker ───────────────────────────────────────────────

class DailyPnlOut(BaseModel):
    trade_date: str        # YYYY-MM-DD in IST
    realized_pnl: Decimal
    open_count: int
    closed_count: int
    circuit_breaker_triggered: bool
    daily_loss_limit_inr: Decimal
    trades_taken_today: int
    max_trades_per_day: int


# ── Trade history ─────────────────────────────────────────────────────────────

class TradeHistoryResponse(BaseModel):
    total: int
    positions: list[PositionOut]


# ── Paper record (the 30-day-clock view) ──────────────────────────────────────

class PaperDayRow(BaseModel):
    date: str              # YYYY-MM-DD in IST
    realized_pnl: Decimal  # NET of charges
    charges: Decimal
    trades: int            # positions closed that day
    profitable: bool       # net realized > 0
    cumulative_pnl: Decimal


class PaperRecordOut(BaseModel):
    """Per-IST-day realized-P&L history for the paper account — the visible
    surface of the 30-day profitable-paper gate (the authoritative gate stays
    Phase 7). P&L is net of trading costs."""

    days: list[PaperDayRow]        # chronological (oldest → newest)
    total_days_traded: int
    profitable_days: int
    losing_days: int
    current_streak: int            # consecutive most-recent profitable days
    best_streak: int
    total_realized_pnl: Decimal    # net
    total_charges: Decimal
    total_trades: int
    win_rate_pct: Decimal
    target_days: int = 30
    start_date: str | None = None
    last_date: str | None = None

File: frontend/src/lib/api/trading.ts
import { api } from './client'

export interface HealthReason {
  code: 'thesis_break' | 'trend_dead' | 'rr_inverted' | 'deep_mae' | 'stale'
  severity: 'watch' | 'cut'
  detail: string
}

export interface PositionHealth {
  verdict: 'hold' | 'watch' | 'cut'
  reasons: HealthReason[]
  drawdown_r: number | null
  rr_remaining: number | null
  regime_er: number | null
}

export interface PositionOut {
  id: string
  user_id: number
  stock_id: number
  symbol: string
  mode: string
  side: 'LONG' | 'SHORT'
  quantity: number
  avg_entry_price: string
  current_sl: string | null
  current_tp: string | null
  trail_state: 'none' | 'breakeven' | 'trailing_1' | 'trailing_2'
  unrealized_pnl: string | null
  realized_pnl: string
  charges: string | null
  exit_price: string | null
  exit_reason: 'sl_hit' | 'tp_hit' | 'manual' | null
  current_price: string | null
  peak_price: string | null
  peak_pnl: string | null
  health: PositionHealth | null
  opened_at: string
  closed_at: string | null
  signal_id: string | null
}

export interface PositionListResponse {
  total: number
  positions: PositionOut[]
}

export interface OrderOut {
  id: string
  user_id: number
  signal_id: string | null
  stock_id: number
  symbol: string
  mode: string
  side: 'BUY' | 'SELL'
  order_type: string
  quantity: number
  price: string | null
  status: string
  placed_at: string
  filled_at: string | null
  filled_price: string | null
  filled_qty: number | null
  error_message: string | null
}

export interface DailyPnlOut {
  trade_date: string
  realized_pnl: string
  total_unrealized_pnl: string
  open_count: number
  closed_count: number
  circuit_breaker_triggered: boolean
  daily_loss_limit_inr: string
  trades_taken_today: number
  max_trades_per_day: number
}

export interface PaperDayRow {
  date: string
  realized_pnl: string
  charges: string
  trades: number
  profitable: boolean
  cumulative_pnl: string
}

export interface PaperRecordOut {
  days: PaperDayRow[]
  total_days_traded: number
  profitable_days: number
  losing_days: number
  current_streak: number
  best_streak: number
  total_realized_pnl: string
  total_charges: string
  total_trades: number
  win_rate_pct: string
  target_days: number
  start_date: string | null
  last_date: string | null
}

// Profit-lock shadow comparator (read-only evidence — controls no orders).
export interface ShadowPolicy {
  policy: string
  exit_price: string | null
  exit_time: string | null
  exit_net: string | null
  still_open: boolean
  capture_pct: number | null
}

export interface ShadowComparison {
  position_id: string
  symbol: string
  side: 'LONG' | 'SHORT'
  quantity: number
  entry: string
  original_sl: string
  classification: string
  bars: number
  peak_price: string | null
  peak_gross: string | null
  actual_exit_price: string | null
  actual_net: string | null
  actual_capture_pct: number | null
  actual_exit_off_tape: boolean
  policies: ShadowPolicy[]
  note: string | null
}

export interface ShadowCompareResponse {
  total: number
  comparisons: ShadowComparison[]
}

export const tradingApi = {
  placeOrder(
    params: { signal_id: string; side?: string; quantity?: number },
    token: string,
  ): Promise<OrderOut> {
    return api.post<OrderOut>('/trading/orders', params, token)
  },

  getOpenPositions(token: string): Promise<PositionListResponse> {
    return api.get<PositionListResponse>('/trading/positions', token)
  },

  closePosition(
    positionId: string,
    exitPrice: string | undefined,
    token: string,
  ): Promise<PositionOut> {
    return api.post<PositionOut>(
      `/trading/positions/${positionId}/close`,
      { exit_price: exitPrice ?? null },
      token,
    )
  },

  updateSl(positionId: string, newSl: string, token: string): Promise<PositionOut> {
    return api.post<PositionOut>(
      `/trading/positions/${positionId}/update-sl`,
      { new_sl: newSl },
      token,
    )
  },

  getHistory(
    params: { limit?: number; offset?: number },
    token: string,
  ): Promise<PositionListResponse> {
    const q = new URLSearchParams()
    if (params.limit != null) q.set('limit', String(params.limit))
    if (params.offset != null) q.set('offset', String(params.offset))
    const qs = q.toString() ? `?${q.toString()}` : ''
    return api.get<PositionListResponse>(`/trading/history${qs}`, token)
  },

  getDailyPnl(token: string): Promise<DailyPnlOut> {
    return api.get<DailyPnlOut>('/trading/daily-pnl', token)
  },

  getShadowCompare(
    params: { limit?: number },
    token: string,
  ): Promise<ShadowCompareResponse> {
    const qs = params.limit != null ? `?limit=${params.limit}` : ''
    return api.get<ShadowCompareResponse>(`/trading/shadow-compare${qs}`, token)
  },

  getPaperRecord(token: string, targetDays = 30): Promise<PaperRecordOut> {
    return api.get<PaperRecordOut>(`/trading/paper-record?target_days=${targetDays}`, token)
  },
}
